README.md
imagetyperz-api-perl - Imagetyperz API wrapper
=========================================

imagetyperzapi is a super easy to use bypass captcha API wrapper for 
imagetyperz.com captcha service

## Installation
    
    git clone https://github.com/imagetyperz-api/imagetyperz-api-perl
     

## How to use?

``` perl
use ImageTyperzAPI;
```

There are 2 ways of authenticating with the server. With 
**access_token** or **username and passord**. 
We encourage you to use the token based authentication because it's more 
secure and username & password authentication might be removed at some 
point from the API libraries.

All the methods that end with **_token** will be used for token 
authentication


**Get balance**

``` perl
ImageTyperzAPI::account_balance_token($access_token);     
```

**Submit image captcha**

``` perl
my $captcha_text = ImageTyperzAPI::solve_captcha_token($access_token, 'captcha.jpg', '1');
```

**Submit recaptcha details**

For recaptcha submission there are two things that are required.
- page_url
- site_key
- type - can be one of this 3 values: `1` - normal, `2` - invisible, `3` - v3 (it's optional, defaults to `1`)
- v3_min_score - minimum score to target for v3 recaptcha `- optional`
- v3_action - action parameter to use for v3 recaptcha `- optional`
- proxy - proxy to use when solving recaptcha, eg. `12.34.56.78:1234` or `12.34.56.78:1234:user:password` `- optional`
- user_agent - useragent to use when solve recaptcha `- optional` 

``` perl
my $recaptcha_params = [
    token      => $access_token,
    #username => $username,       # for legacy auth
    #password   => $password,      # for legacy auth

    action     => 'UPLOADCAPTCHA',
    pageurl    => 'page_url_here',
    googlekey  => 'sitekey_here',

    # v3
    recaptchatype => '3',        # optional, 1 - normal recaptcha, 2 - invisible recaptcha, 3 - v3 recaptcha, default: 1
    captchaaction => 'homepage', # optional, used in solving v3 recaptcha
    score => '0.3',              # optional, min score to target when solving v3 recaptcha

    # proxy
    proxy => '12.34.54.56:123',        # or '123.43.45.65:123:user:password' with auth - optional
    proxytype => 'HTTP', # if proxy is used, un-comment this as well, only HTTP supported for now

    # other optional parameters
    useragent => 'Mozilla/5.0 (X11; Linux x86_64; rv:60.0) Gecko/20100101 Firefox/60.0', # optional
    affiliateid => '12344'    # affiliate id - optional
];
my $captcha_id = ImageTyperzAPI::submit_recaptcha_token($recaptcha_params);
```
This method returns a captchaID. This ID will be used next, to retrieve the g-response, once workers have 
completed the captcha. This takes somewhere between 10-80 seconds.


**Retrieve captcha response**

Once you have the captchaID, you check for it's progress, and later on 
retrieve the gresponse.

The ***in_progress($access_token, $captcha_id)*** method will tell you 
if captcha is still being decoded by workers.
Once it's no longer in progress, you can retrieve the gresponse with 
***retrieve_recaptcha_token($access_token, $captcha_id)***  

``` perl
while(ImageTyperzAPI::in_progress_token($access_token, $captcha_id))
{
    sleep(10);		# sleep for 10 seconds
}

my $gresponse = ImageTyperzAPI::retrieve_recaptcha_token($access_token, $captcha_id);
```

## Other methods/variables

**Legacy auth**

The library has 2 ***sets*** of methods. Those that end with **_token** 
use token auth, and uses the 1st given parameter as token. Methods that 
end with **_legacy** take 2 parameters, username and password, instead 
of access key.

Getting balance with token

```perl
ImageTyperzAPI::account_balance_token($access_token);
```

and here it's the same with legacy/username & password authentication
```perl
ImageTyperzAPI::account_balance_legacy($username, $password);
```

**Affiliate id**

For submitting recaptcha with affiliate ID, set it as the next parameter 
after sitekey.

``` perl
ImageTyperzAPI::submit_recaptcha_token($token, $page_url, $sitekey, $aff_id);
```

Affiliate ID can be set for normal captcha solving as well

**Case sensitive**

Regular captcha takes a case-sensitive argument, after the image 
parameter, which tells the server if captcha is case sensitive or not. 1 
means sensitive.
``` perl
ImageTyperzAPI::solve_captcha_token($access_token, 'captcha.jpg', '1');
```

There's a 4th parameter which can be given here, which is the affiliate 
id, after caseSensitive.


**Get details of proxy for recaptcha**

In case you submitted the recaptcha with proxy, you can check the status of the proxy, if it was used or not,
and if not, what the reason was with the following:

``` perl
printf 'Was proxy used: %s\n', ImageTyperzAPI::was_proxy_used_token($access_token, $captcha_id);
```

The response is in JSON and looks like this:
```json
{
    "Result": "gresponse from solving, or empty if not solved yet",
    "Proxy_client": "proxy submitted by client (if any)",
    "Proxy_worker": "proxy used by worker, in case client submitted, and no errors with proxy",
    "Proxy_reason": "in case of proxy not working, reason will be found here"
}
```

**Set captcha bad**

When a captcha was solved wrong by our workers, you can notify the 
server with it's ID,
so we know something went wrong.

``` perl
ImageTyperzAPI::set_captcha_bad_token($access_token, $captcha_id);
```

## Examples
Check example.pl

## License
API library is licensed under the MIT License

## More information
More details about the server-side API can be found [here](http://imagetyperz.com)


<sup><sub>captcha, bypasscaptcha, decaptcher, decaptcha, 2captcha, deathbycaptcha, anticaptcha, 
bypassrecaptchav2, bypassnocaptcharecaptcha, bypassinvisiblerecaptcha, captchaservicesforrecaptchav2, 
recaptchav2captchasolver, googlerecaptchasolver, recaptchasolverpython, recaptchabypassscript</sup></sub>



changelog.md
18.03.2019
------
- added geetest captcha


lib/ImageTyperzAPI.pm
#!/bin/perl

package ImageTyperzAPI;

use warnings;
use strict;
use LWP::UserAgent;
use HTTP::Request::Common;
use MIME::Base64 qw(encode_base64);
use URI qw( );

# constants
# --------------------------------------------------------------------------------------------
my $CAPTCHA_ENDPOINT = 'http://captchatypers.com/Forms/UploadFileAndGetTextNEW.ashx';
my $RECAPTCHA_SUBMIT_ENDPOINT = 'http://captchatypers.com/captchaapi/UploadRecaptchaV1.ashx';
my $RECAPTCHA_RETRIEVE_ENDPOINT = 'http://captchatypers.com/captchaapi/GetRecaptchaText.ashx';
my $BALANCE_ENDPOINT = 'http://captchatypers.com/Forms/RequestBalance.ashx';
my $BAD_IMAGE_ENDPOINT = 'http://captchatypers.com/Forms/SetBadImage.ashx';
my $PROXY_CHECK_ENDPOINT = 'http://captchatypers.com/captchaAPI/GetReCaptchaTextJSON.ashx';
my $GEETEST_SUBMIT_ENDPOINT = 'http://captchatypers.com/captchaapi/UploadGeeTest.ashx';
my $GEETEST_RETRIEVE_ENDPOINT = 'http://captchatypers.com/captchaapi/getrecaptchatext.ashx';

my $CAPTCHA_ENDPOINT_CONTENT_TOKEN = 'http://captchatypers.com/Forms/UploadFileAndGetTextNEWToken.ashx';
my $CAPTCHA_ENDPOINT_URL_TOKEN = 'http://captchatypers.com/Forms/FileUploadAndGetTextCaptchaURLToken.ashx';
my $RECAPTCHA_SUBMIT_ENDPOINT_TOKEN = 'http://captchatypers.com/captchaapi/UploadRecaptchaToken.ashx';
my $RECAPTCHA_RETRIEVE_ENDPOINT_TOKEN = 'http://captchatypers.com/captchaapi/GetRecaptchaTextToken.ashx';
my $BALANCE_ENDPOINT_TOKEN = 'http://captchatypers.com/Forms/RequestBalanceToken.ashx';
my $BAD_IMAGE_ENDPOINT_TOKEN = 'http://captchatypers.com/Forms/SetBadImageToken.ashx';
my $PROXY_CHECK_ENDPOINT_TOKEN = 'http://captchatypers.com/captchaAPI/GetReCaptchaTextTokenJSON.ashx';
my $GEETEST_SUBMIT_ENDPOINT_TOKEN = 'http://captchatypers.com/captchaapi/UploadGeeTestToken.ashx';

# ACCESS TOKEN
# --------------
# Solve normal captcha
# --------------------------------------------------
sub solve_captcha_token
{
	my $ref_id = '0';
	my $chkcase = '0';
	
	my $ua = LWP::UserAgent->new();
	if(defined $_[2])		# check if chkcase was given
	{
		$chkcase = $_[2];	# set chkcase
	}
	if(defined $_[3])		# check if ref id was given
	{
		$ref_id = $_[3];	# set ref id
	}

	# read file
    local $/ = undef;
    open FILE, $_[1] or die "Couldn't open file: $!";
    my $string = <FILE>;
    close FILE;

	my $response = $ua->request(POST $CAPTCHA_ENDPOINT_CONTENT_TOKEN, Content => [
				 action => 'UPLOADCAPTCHA',
				 token => $_[0],
				 file => encode_base64($string),
				 chkCase => $chkcase,
				 affiliateid => $ref_id
				]);

   	if ($response->is_error())
   	{
		return $response->status_line;
    } else {
		my $c = $response->content();
		if (index($c, 'ERROR') != -1) {
			die($c);
		} 
		return replace("Uploading file...", "", $c);
    }
}

# Submit recaptcha
# -------------------------------------------------------------------------------------------------
sub submit_recaptcha_token
{
	my $ua = LWP::UserAgent->new();
	my $data = $_[0];
	
	my $response = $ua->request(POST $RECAPTCHA_SUBMIT_ENDPOINT_TOKEN, Content_Type => 'form-data', Content => $data);

   	if ($response->is_error())
   	{
		return $response->status_line;
    } else {
		my $c = $response->content();
		if (index($c, 'ERROR') != -1) {
			die($c);
		} 
        return $c;		# return ID
    }
}

# Retrieve recaptcha
# -------------------------------------------------------------------------------------------------
sub retrieve_recaptcha_token
{
	my $ua = LWP::UserAgent->new();
	my $response = $ua->request(POST $RECAPTCHA_RETRIEVE_ENDPOINT_TOKEN, Content_Type => 'form-data', Content => [
				 action => 'GETTEXT',
				 token => $_[0],
				 captchaid => $_[1],
				]);

   	if ($response->is_error())
   	{
		return $response->status_line;
    } else {
		my $c = $response->content();
		if (index($c, 'ERROR') != -1) {
			if (index($c, 'NOT_DECODED') != -1) {		
				 return $c;		# return NOT_DECODED
			}
			else  {die($c);}	# error, die
		} 
        return $c;		# return ID	
    }
}

# Checks if recaptcha is still in process of solving
sub in_progress_token
{
	my $resp = retrieve_recaptcha_token($_[0], $_[1]);
	if(index($resp, 'NOT_DECODED') == -1)
	{
		return 0;		# does not contain NOT_DECODED, move on
	}
	else
	{
		return 1;		# contains NOT_DECODED, still in progress
	}
}

# Checks if geetest is still in process of solving
sub in_progress_geetest
{
	my $resp = retrieve_geetest($_[0]);
	if(index($resp, 'NOT_DECODED') == -1)
	{
		return 0;		# does not contain NOT_DECODED, move on
	}
	else
	{
		return 1;		# contains NOT_DECODED, still in progress
	}
}

# Submit geetest
sub submit_geetest_token
{
	my $ua = LWP::UserAgent->new();
	my $data = $_[0];

	my $url = $GEETEST_SUBMIT_ENDPOINT_TOKEN;
	$url .= '?';
	my $u = URI->new('', 'http');
	$u->query_form(@$data);
	my $query = $u->query;
	$url .= $query;
	my $response = $ua->request(GET $url);

	if ($response->is_error())
	{
		return $response->status_line;
	} else {
		my $c = $response->content();
		if (index($c, 'ERROR') != -1) {
			die($c);
		}
		return $c;		# return ID
	}
}

# Check account balance
# -------------------------------------------------------------------------------------------------
sub account_balance_token
{
	my $ua = LWP::UserAgent->new();
	my $response = $ua->request(POST $BALANCE_ENDPOINT_TOKEN, Content_Type => 'form-data', Content => [
				 action => 'REQUESTBALANCE',
				 token => $_[0],
				 'submit' => 'Submit'
				]);

    if ($response->is_error())
    {
		return $response->status_line;			# return error
    } else {
		my $c = $response->content();
		if (index($c, 'ERROR') != -1) {
			die($c);
		} 
        return '$' . $c;		# return balance
    }
}

# Set captcha as BAD
sub set_captcha_bad_token
{
	my $ua = LWP::UserAgent->new();
	my $response = $ua->request(POST $BAD_IMAGE_ENDPOINT_TOKEN, Content_Type => 'form-data', Content => [
				 action => 'SETBADIMAGE',
				 token => $_[0],
				 imageid =>$_[1],
				 submit => "Submissssst"
				]);

   	if ($response->is_error())
   	{
		return $response->status_line;
    } else {
        my $c = $response->content();
		if (index($c, 'ERROR') != -1) {
			die($c);
		} 
        return $c;		
    }
}

# Tells if proxy was used, and if not reason why - token
sub was_proxy_used_token
{
	my $ua = LWP::UserAgent->new();
	my $response = $ua->request(POST $PROXY_CHECK_ENDPOINT_TOKEN, Content_Type => 'form-data', Content => [
		action => 'GETTEXT',
		token => $_[0],
		captchaid =>$_[1],
	]);

	if ($response->is_error())
	{
		return $response->status_line;
	} else {
		my $c = $response->content();
		if (index($c, 'Error') != -1) {
			die($c);
		}
		return $c;
	}
}

# Tells if proxy was used, and if not reason why - legacy
sub was_proxy_used_legacy
{
	my $ua = LWP::UserAgent->new();
	my $response = $ua->request(POST $PROXY_CHECK_ENDPOINT, Content_Type => 'form-data', Content => [
		action => 'GETTEXT',
		username => $_[0],
		password => $_[1],
		captchaid =>$_[2],
	]);

	if ($response->is_error())
	{
		return $response->status_line;
	} else {
		my $c = $response->content();
		if (index($c, 'Error') != -1) {
			die($c);
		}
		return $c;
	}
}


# LEGACY WAY
# this might get deprecated, better of using access_token
# --------------------------------------------------------

# Solve normal captcha
# --------------------------------------------------
sub solve_captcha_legacy
{
	my $ref_id = '0';
	my $chkcase = '0';
	
	my $ua = LWP::UserAgent->new();
	if(defined $_[3])		# check if chkcase was given
	{
		$chkcase = $_[3];	# set chkcase
	}
	if(defined $_[4])		# check if ref id was given
	{
		$ref_id = $_[4];	# set ref id
	}

	# read file
    local $/ = undef;
    open FILE, $_[2] or die "Couldn't open file: $!";
    my $string = <FILE>;
    close FILE;

	my $response = $ua->request(POST $CAPTCHA_ENDPOINT, Content => [
				 action => 'UPLOADCAPTCHA',
				 username => $_[0],
				 password => $_[1],
				 file => encode_base64($string),
				 chkCase => $chkcase,
				 affiliateid => $ref_id
				]);

   	if ($response->is_error())
   	{
		return $response->status_line;
    } else {
		my $c = $response->content();
		if (index($c, 'ERROR') != -1) {
			die($c);
		} 
		return replace("Uploading file...", "", $c);
    }
}

# Submit recaptcha
# -------------------------------------------------------------------------------------------------
sub submit_recaptcha_legacy
{
	my $ua = LWP::UserAgent->new();
	my $data = $_[0];

	my $response = $ua->request(POST $RECAPTCHA_SUBMIT_ENDPOINT, Content_Type => 'form-data', Content => $data);

	if ($response->is_error())
	{
		return $response->status_line;
	} else {
		my $c = $response->content();
		if (index($c, 'ERROR') != -1) {
			die($c);
		}
		return $c;		# return ID
	}
}

# Retrieve recaptcha
# -------------------------------------------------------------------------------------------------
sub retrieve_recaptcha_legacy
{
	my $ua = LWP::UserAgent->new();
	my $response = $ua->request(POST $RECAPTCHA_RETRIEVE_ENDPOINT, Content_Type => 'form-data', Content => [
				 action => 'GETTEXT',
				 username => $_[0],
				 password => $_[1],
				 captchaid => $_[2],
				]);

   	if ($response->is_error())
   	{
		return $response->status_line;
    } else {
		my $c = $response->content();
		if (index($c, 'ERROR') != -1) {
			if (index($c, 'NOT_DECODED') != -1) {		
				 return $c;		# return NOT_DECODED
			}
			else {die($c);}	# error, die
		} 
        return $c;		# return ID	
    }
}

# Retrieve recaptcha
# -------------------------------------------------------------------------------------------------
sub retrieve_geetest
{
	my $data = $_[0];
	my $ua = LWP::UserAgent->new();
	my $url = $GEETEST_RETRIEVE_ENDPOINT;
	$url .= '?';
	my $u = URI->new('', 'http');
	$u->query_form(@$data);
	my $query = $u->query;
	$url .= $query;
	my $response = $ua->request(GET $url);

	if ($response->is_error())
	{
		return $response->status_line;
	} else {
		my $c = $response->content();
		if (index($c, 'ERROR') != -1) {
			if (index($c, 'NOT_DECODED') != -1) {
				return $c;		# return NOT_DECODED
			}
			else {die($c);}	# error, die
		}
		return $c;		# return ID
	}
}

# Checks if recaptcha is still in process of solving
sub in_progress_legacy
{
	my $resp = retrieve_recaptcha_legacy($_[0], $_[1], $_[2]);
	if(index($resp, 'NOT_DECODED') == -1)
	{
		return 0;		# does not contain NOT_DECODED, move on
	}
	else
	{
		return 1;		# contains NOT_DECODED, still in progress
	}
}

# Submit geetest
sub submit_geetest_legacy
{
	my $ua = LWP::UserAgent->new();
	my $data = $_[0];

	my $url = $GEETEST_SUBMIT_ENDPOINT;
	$url .= '?';
	my $u = URI->new('', 'http');
	$u->query_form(@$data);
	my $query = $u->query;
	$url .= $query;
	my $response = $ua->request(GET $url);

	if ($response->is_error())
	{
		return $response->status_line;
	} else {
		my $c = $response->content();
		if (index($c, 'ERROR') != -1) {
			die($c);
		}
		return $c;		# return ID
	}
}

# Check account balance
# -------------------------------------------------------------------------------------------------
sub account_balance_legacy
{
	my $ua = LWP::UserAgent->new();
	my $response = $ua->request(POST $BALANCE_ENDPOINT, Content_Type => 'form-data', Content => [
				 action => 'REQUESTBALANCE',
				 username => $_[0],
				 password => $_[1],
				 'submit' => 'Submit'
				]);

    if ($response->is_error())
    {
		return $response->status_line;			# return error
    } else {
		my $c = $response->content();
		if (index($c, 'ERROR') != -1) {
			die($c);
		} 
        return '$' . $c;	
    }
}

# Set captcha as BAD
sub set_captcha_bad_legacy
{
	my $ua = LWP::UserAgent->new();
	my $response = $ua->request(POST $BAD_IMAGE_ENDPOINT, Content_Type => 'form-data', Content => [
				 action => 'SETBADIMAGE',
				 username => $_[0],
				 password => $_[1],
				 imageid =>$_[2],
				 submit => "Submissssst"
				]);

   	if ($response->is_error())
   	{
		return $response->status_line;
    } else {
        my $c = $response->content();
		if (index($c, 'ERROR') != -1) {
			die($c);
		} 
        return $c;		
    }
}

# replace string
sub replace {
	  my ($from,$to,$string) = @_;
	  $string =~s/$from/$to/ig;                          #case-insensitive/global (all occurrences)

	  return $string;
}

1;


example.pl
#!/bin/perl

use warnings;
use strict;
use File::Basename;
use File::Spec;
# set library path
use lib File::Spec->catfile(dirname(__FILE__), 'lib');
use ImageTyperzAPI;

sub test_api {
    # access token, get from: http://www.imagetyperz.com/Forms/ClientHome.aspx
    my $access_token = 'access_token_here';

    # set your own username and password, legacy way, use access_token instead
    # ----------------------------------
    my $username = 'your_username';     # legacy authentication, will be deprecated at some point
    my $password = 'your_password';     # legacy

    # check account balance
    # ----------------------
    #printf 'Balance token: %s\n', ImageTyperzAPI::account_balance_token($access_token);
    printf 'Balance legacy: %s\n', ImageTyperzAPI::account_balance_legacy($username, $password);
    # solve normal captcha
    # --------------------
    my $captcha_text = ImageTyperzAPI::solve_captcha_token($access_token, 'captcha.jpg', '1');
    printf 'Captcha text: %s\n', $captcha_text;
    # ==============================================================================
    # solve recaptcha
    # -------------------------
    # submit
    # -------
    my $recaptcha_params = [
        token      => $access_token,
        #username => $username,       # for legacy auth
        #password   => $password,      # for legacy auth

        action     => 'UPLOADCAPTCHA',
        pageurl    => 'page_url_here',
        googlekey  => 'sitekey_here',

        # v3
        # recaptchatype => '3',        # optional, 1 - normal recaptcha, 2 - invisible recaptcha, 3 - v3 recaptcha, default: 1
        # captchaaction => 'homepage', # optional, used in solving v3 recaptcha
        # score => '0.3',              # optional, min score to target when solving v3 recaptcha

        # proxy
        # proxy => '12.34.54.56:123',        # or '123.43.45.65:123:user:password' with auth - optional
        # proxytype => 'HTTP', # if proxy is used, un-comment this as well, only HTTP supported for now

        # other optional parameters
        # useragent => 'Mozilla/5.0 (X11; Linux x86_64; rv:60.0) Gecko/20100101 Firefox/60.0', # optional
        # affiliateid => '12344'    # affiliate id - optional
    ];
    my $captcha_id = ImageTyperzAPI::submit_recaptcha_token($recaptcha_params);
    #my $captcha_id = ImageTyperzAPI::submit_recaptcha_legacy($recaptcha_params);
    printf 'Recaptcha ID: %s', $captcha_id;
    # retrieve
    # -------
    while (ImageTyperzAPI::in_progress_token($access_token, $captcha_id)) # while in progress
    #while(ImageTyperzAPI::in_progress_legacy($username, $password, $captcha_id))	# while in progress
    {
        sleep(10); # sleep for 10 seconds
    }
    printf 'Recaptcha response token: %s\n', ImageTyperzAPI::retrieve_recaptcha_token($access_token, $captcha_id);
    #printf 'Recaptcha response token: %s\n', ImageTyperzAPI::retrieve_recaptcha_legacy($username, $password, $captcha_id);

    #recaptcha
    #----------
    #while(ImageTyperzAPI::in_progress_legacy($username, $password, $captcha_id2))	# while in progress
    #{
    #	sleep(10);		# sleep for 10 seconds
    #}
    #printf 'Recaptcha response legacy: %s\n', ImageTyperzAPI::retrieve_recaptcha_legacy($username, $password, $captcha_id2);

    ## Geetest
    my $geetest_params = [(
        token     => $access_token,
        username  => $username, # legacy
        password  => $password, # legacy

        action    => 'UPLOADCAPTCHA',
        domain    => 'domain',
        challenge => 'geetest challenge',
        gt        => 'geetest gt'

        # proxy
        # proxy => '12.34.54.56:123',        # or '123.43.45.65:123:user:password' with auth - optional
        # proxytype => 'HTTP', # if proxy is used, un-comment this as well, only HTTP supported for now

        # other optional parameters
        # useragent => 'Mozilla/5.0 (X11; Linux x86_64; rv:60.0) Gecko/20100101 Firefox/60.0', # optional
        # affiliateid => '12344'    # affiliate id - optional
    )];
    #my $geetest_id = ImageTyperzAPI::submit_geetest_token($geetest_params);
    my $geetest_id = ImageTyperzAPI::submit_geetest_legacy($geetest_params);
    printf 'Geetest ID: %s', $geetest_id;

    while(ImageTyperzAPI::in_progress_geetest([(
        token     => $access_token,
        username  => $username,     # legacy
        password  => $password,     # legacy
        captchaid => $geetest_id,
        action    => 'GETTEXT'
    )]))	# while in progress
    {
    	sleep(10);		# sleep for 10 seconds
    }
    printf 'Geetest response: %s\n', ImageTyperzAPI::retrieve_geetest([(
        token     => $access_token,
        username  => $username, # legacy
        password  => $password, # legacy
        captchaid => $geetest_id,
        action    => 'GETTEXT'
    )]);
    # challenge;;;validate;;;seccode
    # 5fcc6e26a128024f7327a355350043007t;;;cb9d32041e4765e9f825a0fa03e3ebeb;;;cb9d32041e4765e9f825a0fa03e3ebeb|jordan


    # Other examples
    # ---------------------------------------------------------------------------------------------------------------------------------
    # my $captcha_text = ImageTyperzAPI::solve_captcha_legacy($username, $password, 'captcha.jpg', '1', $ref_id);
    # my $captcha_text = ImageTyperzAPI::solve_captcha_token($access_token, 'captcha.jpg', '1', $ref_id);	# with caseSensitive and affiliate_id

    # method to check if proxy was used, response is json
    # {
    #     "Result": "gresponse from solving, or empty if not solved yet",
    #     "Proxy_client": "proxy submitted by client (if any)",
    #     "Proxy_worker": "proxy used by worker, in case client submitted, and no errors with proxy",
    #     "Proxy_reason": "in case of proxy not working, reason will be found here"
    # }
    #printf 'Was proxy used: %s\n', ImageTyperzAPI::was_proxy_used_token($access_token, $captcha_id);
    #printf 'Was proxy used: %s\n', ImageTyperzAPI::was_proxy_used_legacy($username, $password, $captcha_id);

    # printf 'Set captcha bad response token: %s\n', ImageTyperzAPI::set_captcha_bad_token($access_token, $captcha_id);		# set captcha as bad
    # printf 'Set captcha bad response legacy: %s\n', ImageTyperzAPI::set_captcha_bad_legacy($username, $password, $captcha_id2);		# set captcha as bad, legacy
}

test_api(); # test API

1;
